sql/core/src/main/scala/org/apache/spark/sql/SparkSession.scala

```diff
@@ -776,6 +779,8 @@ class SparkSession private(
 @Stable
 object SparkSession extends Logging {
 
+  private[spark] val numActiveSessions: AtomicInteger = new AtomicInteger(0)
```

Comment:
This is error prone and also not easy to debug, we may need to keep track of each active SparkSession in SparkContext. We don't need to pass in the SparkSession instance, maybe only compute the `System.identityHashCode()` of current SparkSession object and pass that to the SparkContext. On SparkSession.stop() we only need to drop the corresponding identity from SparkContext, and when it's empty it means we can safely stop the SparkContext, too.

Reply:
Yeah, I think the identity hash code could be an interesting solution here. I'll experiment with that